### PR TITLE
feat(IOS-F-05): IMAP Client — TLS connection, XOAUTH2 auth, IDLE, connection pooling

### PR DESCRIPTION
**Pull request by vivekjyani**

### CODE REVIEW

File: PrivateMailPackage/Sources/PrivateMailFeature/Data/Network/ConnectionPool.swift [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        }
+
+        // Second: create a new connection if under the limit
+        guard entries.count < maxConnectionsPerAccount else {
+            throw IMAPError.commandFailed(
+                "Connection pool exhausted: \(entries.count)/\(maxConnectionsPerAccount) for account \(accountId)"
+            )
+        }
+
+        let client = IMAPClient()
```
Comment (VishnuBishnoi) [inline PrivateMailPackage/Sources/PrivateMailFeature/Data/Network/ConnectionPool.swift]: **Critical — Spec Violation (FR-SYNC-09)**

FR-SYNC-09 states:
> "Connection pool exhaustion **MUST** queue operations until a connection becomes available, not open additional connections beyond the limit."

The current implementation throws immediately when the pool is full. This should instead suspend the caller until a connection is checked back in — e.g., using `AsyncStream`, a continuation queue, or a semaphore-style actor pattern.

Additionally, `commandFailed` is a misleading error type here. If keeping the throw as a fallback/timeout, consider adding a dedicated `poolExhausted` case to `IMAPError`.

```swift
// Suggested approach: maintain a waiters queue
private var waiters: [String: [CheckedContinuation<IMAPClient, Error>]] = [:]

// In checkout, when pool is full:
return try await withCheckedThrowingContinuation { cont in
    waiters[accountId, default: []].append(cont)
}

// In checkin, resume the first waiter:
if let waiter = waiters[accountId]?.removeFirst() {
    waiter.resume(returning: client)
} else {
    entries[i].isCheckedOut = false
}
```

File: PrivateMailPackage/Sources/PrivateMailFeature/Data/Network/IMAPClient.swift [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+                    try await session.stopIDLE()
+                    _ = try await session.startIDLE()
+                    lastIDLEStart = Date()
+                    continue
+                }
+
+                let notification = try await session.readIDLENotification()
+
+                // Check for new mail notification: "* <N> EXISTS"
+                if notification.contains("EXISTS") {
```
Comment (VishnuBishnoi) [inline PrivateMailPackage/Sources/PrivateMailFeature/Data/Network/IMAPClient.swift]: **Important — IDLE re-issue will miss the 25-minute window**

The time check happens *before* `readIDLENotification()`, which blocks indefinitely waiting for server data. If no notification arrives, the read will block past the 25-minute window, and Gmail will drop the connection at ~29 minutes.

This needs a timeout-based approach — either:
1. Race `readIDLENotification()` against a `Task.sleep` using a task group
2. Add a read timeout to `IMAPSession.readIDLENotification()` that fires before the 25-minute mark

```swift
// Example using task group:
try await withThrowingTaskGroup(of: String?.self) { group in
    group.addTask { try await self.session.readIDLENotification() }
    group.addTask {
        let remaining = refreshInterval - Date().timeIntervalSince(lastIDLEStart)
        try await Task.sleep(for: .seconds(max(remaining, 0)))
        return nil // signal re-issue
    }
    // first result wins
}
```

File: PrivateMailPackage/Sources/PrivateMailFeature/Data/Network/IMAPSession.swift [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+        guard let conn = connection, conn.state == .ready else {
+            throw IMAPError.connectionFailed("Not connected")
+        }
+
+        let data: Data = try await withTaskCancellationHandler {
+            try await withCheckedThrowingContinuation { cont in
+                conn.receive(minimumIncompleteLength: 1, maximumLength: 65536) { data, _, _, error in
+                    if let error {
+                        cont.resume(throwing: IMAPError.connectionFailed(error.localizedDescription))
+                    } else if let data, !data.isEmpty {
```
Comment (VishnuBishnoi) [inline PrivateMailPackage/Sources/PrivateMailFeature/Data/Network/IMAPSession.swift]: **Important — No receive timeout**

`receiveMoreData()` will block indefinitely if the server stops responding. This means any IMAP command (not just IDLE) can hang forever on a dead connection.

Consider adding a timeout using the same `DispatchQueue.asyncAfter` + `AtomicFlag` pattern already used in `connect()`, or race `Task.sleep` against the receive.

Without this, a network partition during sync will cause the entire sync to hang rather than fail and retry per FR-SYNC-09.

File: PrivateMailPackage/Sources/PrivateMailFeature/Data/Network/IMAPResponseParser.swift [hunk 4]
```diff
@@ -0,0 +1,728 @@
...
+    /// - RFC 2822: "Mon, 1 Jan 2024 12:00:00 +0000"
+    /// - IMAP internal: "01-Jan-2024 12:00:00 +0000"
+    private static func parseIMAPDate(_ raw: String?) -> Date? {
+        guard let raw, !raw.isEmpty else { return nil }
+        let trimmed = raw.trimmingCharacters(in: .whitespaces)
+
+        let formatters: [DateFormatter] = {
+            let f1 = DateFormatter()
+            f1.dateFormat = "EEE, d MMM yyyy HH:mm:ss Z"
+            f1.locale = Locale(identifier: "en_US_POSIX")
```
Comment (VishnuBishnoi) [inline PrivateMailPackage/Sources/PrivateMailFeature/Data/Network/IMAPResponseParser.swift]: **Important — Performance: DateFormatter allocation per call**

This closure creates 3 `DateFormatter` instances on every invocation. `DateFormatter` is expensive to allocate (~10-50µs each). During a full sync of 1,000 emails, this creates 3,000+ formatters unnecessarily.

These should be `static let` properties on the enum:

```swift
private static let dateFormatters: [DateFormatter] = {
    let f1 = DateFormatter()
    f1.dateFormat = "EEE, d MMM yyyy HH:mm:ss Z"
    f1.locale = Locale(identifier: "en_US_POSIX")
    // ... etc
    return [f1, f2, f3]
}()
```

File: PrivateMailPackage/Sources/PrivateMailFeature/Data/Network/IMAPClient.swift [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+    /// Per FR-SYNC-08: MUST extract attachment metadata from BODYSTRUCTURE.
+    public func fetchBodies(uids: [UInt32]) async throws -> [IMAPEmailBody] {
+        guard !uids.isEmpty else { return [] }
+
+        var bodies: [IMAPEmailBody] = []
+
+        for uid in uids {
+            // Step 1: Fetch BODYSTRUCTURE
+            let bsCommand = "UID FETCH \(uid) (UID BODYSTRUCTURE)"
+            let bsResponses = try await session.execute(bsCommand)
```
Comment (VishnuBishnoi) [inline PrivateMailPackage/Sources/PrivateMailFeature/Data/Network/IMAPClient.swift]: **Medium — Sequential body fetch is an N+1 problem**

This loop issues 1-3 IMAP commands per UID (BODYSTRUCTURE + body parts). For 100 emails, that is 200-300 individual round trips.

Consider:
1. Batching the BODYSTRUCTURE fetch: `UID FETCH 101,102,...,200 (UID BODYSTRUCTURE)`
2. Then batching body part fetches for UIDs that share the same structure

This will significantly impact NFR-SYNC-02 (initial sync < 60s for 1,000 emails).

File: PrivateMailPackage/Sources/PrivateMailFeature/Data/Network/IMAPClient.swift [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+    }
+
+    /// Selects a folder for subsequent operations.
+    ///
+    /// Returns UIDVALIDITY and EXISTS count from the server's SELECT response.
+    public func selectFolder(_ imapPath: String) async throws -> (uidValidity: UInt32, messageCount: UInt32) {
+        let responses = try await session.execute("SELECT \"\(imapPath)\"")
+
+        let uidValidity = IMAPResponseParser.parseUIDValidity(from: responses)
+        let messageCount = IMAPResponseParser.parseExists(from: responses)
```
Comment (VishnuBishnoi) [inline PrivateMailPackage/Sources/PrivateMailFeature/Data/Network/IMAPClient.swift]: **Medium — IMAP command injection risk**

Folder paths are interpolated directly into IMAP commands without sanitization. A folder name containing `"` or CRLF (`\r\n`) could break the command framing or inject additional commands.

This applies to `selectFolder`, `copyMessages`, and `appendMessage`. Consider adding input validation or escaping for all values interpolated into IMAP command strings.

File: PrivateMailPackage/Sources/PrivateMailFeature/Data/Sync/GmailFolderMapper.swift [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        // \All (Gmail All Mail) is redundant
+        if lowered.contains("\\all") {
+            return false
+        }
+
+        // Gmail Important is a flag, not a real folder
+        if imapPath == "[Gmail]/Important" {
+            return false
+        }
+
```
Comment (VishnuBishnoi) [inline PrivateMailPackage/Sources/PrivateMailFeature/Data/Sync/GmailFolderMapper.swift]: **Medium — `\Important` attribute not checked**

Per FR-SYNC-01, `\Important` **MUST NOT** be synced. This check only matches the exact path `[Gmail]/Important`, but localized Gmail servers may use a different path. The `\Important` attribute should also be checked, similar to how `\All` is checked on line 87:

```swift
if lowered.contains("\\important") {
    return false
}
```

**Review by VishnuBishnoi — CHANGES_REQUESTED** [review]
## PR Review: IMAP Client Layer

### Overall Assessment

Strong architecture with clean Domain/Data separation (AI-01/AI-02), proper actor isolation, zero external dependencies (TC-03, P-02), and excellent spec traceability. Test coverage is comprehensive with good edge case handling.

However, there is one spec violation that should be fixed before merge, plus several issues that will affect production reliability and performance.

### Summary

| Category | Count |
|----------|-------|
| Critical | 1 (pool queueing spec violation) |
| Important | 3 (IDLE timing, receive timeout, DateFormatter perf) |
| Medium | 4 (N+1 fetch, injection, `\Important`, batch STORE) |
| Low | 3 (recursion, queue label, mock sendability) |
